File: dotnet/GHIElectronics.DUELink/Rtc.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net;
using System.Text;
using System.Threading.Tasks;
using static System.Runtime.InteropServices.JavaScript.JSType;

namespace GHIElectronics.DUELink {
    public partial class DUELinkController {
        public class RtcController {
            SerialInterface serialPort = null;
            StreamController stream;

            public RtcController(SerialInterface serialPort, StreamController stream) {
                this.serialPort = serialPort;
                this.stream = stream;
            }

            public bool Write(byte[] rtc_timedate) {

                //var write_array = string.Empty;

                //write_array = "[";

                //for (var i = 0; i < timedate.Length; i++) {
                //    write_array += timedate[i];

                //    if (i < timedate.Length - 1)
                //        write_array += ",";
                //}

                //write_array += "]";

                //var cmd = string.Format("RtcW({0})", write_array);

                var cmd = $"dim b9[{rtc_timedate.Length}]";
                this.serialPort.WriteCommand(cmd);
                this.serialPort.ReadResponse();

                var written = this.stream.WriteBytes("b9", rtc_timedate);

                cmd = "RtcW(b9)";
                this.serialPort.WriteCommand(cmd);

                var ret = this.serialPort.ReadResponse();

                return ret.success;                
                               
            }

            public int Read(byte[] rtc_timedate) {
                var cmd = $"dim b9[{rtc_timedate.Length}]";
                this.serialPort.WriteCommand(cmd);
                this.serialPort.ReadResponse();

                // we can't check response as Asio(1) there will be no response                                              
                this.serialPort.WriteCommand("RtcR(b9)");

                this.serialPort.ReadResponse();

                // use stream to read b9
                var ret = this.stream.ReadBytes("b9", rtc_timedate);

                return ret;
            }

            //public string ReadFmt() {
            //    this.serialPort.WriteCommand("RtcR(0)");

            //    var ret = this.serialPort.ReadResponse();

            //    return ret.response;

            //}
            
        }
    }
}

File: python/DUELink/Rtc.py
from enum import Enum
from DUELink.SerialInterface import SerialInterface
from DUELink.Stream import StreamController

class RtcController:   

    def __init__(self, serialPort:SerialInterface, stream:StreamController):
        self.serialPort = serialPort
        self.stream = stream

    def RtcW(self, rtc_timedate: bytes)->bool:
        count = len(rtc_timedate)
        # declare b9 array
        cmd = f"dim b9[{count}]"
        self.serialPort.WriteCommand(cmd)
        self.serialPort.ReadResponse()

        # write data to b9
        ret = self.stream.WriteBytes("b9",rtc_timedate)

        # write b9 to dmx
        self.serialPort.WriteCommand("RtcW(b9)")
        ret = self.serialPort.ReadResponse()

        return ret.success

    def RtcR(self, rtc_timedate: bytearray)->int:
        count = len(rtc_timedate)
        # declare b9 array
        cmd = f"dim b9[{count}]"
        self.serialPort.WriteCommand(cmd)
        self.serialPort.ReadResponse()

        cmd = f"RtcR(b9)"
        self.serialPort.WriteCommand(cmd)
        self.serialPort.ReadResponse()

        ret = self.stream.ReadBytes("b9",rtc_timedate)

        return ret
    
    def RtcShow(self)->bool:
        self.serialPort.WriteCommand("OtpR(0)")
        ret = self.serialPort.ReadResponse()

        return ret.success
        
        




       



